src/include/optimizer/rule_impls.h

```diff
@@ -0,0 +1,823 @@
+   * @param input Input OperatorExpression to transform
+   * @param transformed Vector of transformed OperatorExpressions
+   * @param context Current OptimizeContext executing under
+   */
+  void Transform(common::ManagedPointer<OperatorExpression> input,
+                 std::vector<std::unique_ptr<OperatorExpression>> *transformed,
+                 OptimizeContext *context) const override;
+};
+
+/**
+ * Rule performs predicate push-down to push a filter through aggregation, also
+ * will embed filter into aggregation operator if appropriate.
+ */
+class PushFilterThroughAggregation : public Rule {
+ public:
+  /**
+   * Constructor
+   */
+  PushFilterThroughAggregation();
+
+  /**
+   * Checks whether the given rule can be applied
+   * @param plan OperatorExpression to check
+   * @param context Current OptimizeContext executing under
+   * @returns Whether the input OperatorExpression passes the check
+   */
+  bool Check(common::ManagedPointer<OperatorExpression> plan, OptimizeContext *context) const override;
+
+  /**
+   * Transforms the input expression using the given rule
+   * @param input Input OperatorExpression to transform
+   * @param transformed Vector of transformed OperatorExpressions
+   * @param context Current OptimizeContext executing under
+   */
+  void Transform(common::ManagedPointer<OperatorExpression> input,
+                 std::vector<std::unique_ptr<OperatorExpression>> *transformed,
+                 OptimizeContext *context) const override;
+};
+
+/**
+ * Rule embeds a filter into a scan operator. After predicate push-down, we
+ * eliminate all filters in the operator trees. Predicates should be associated
+ * with get or join
+ */
+class EmbedFilterIntoGet : public Rule {
+ public:
+  /**
+   * Constructor
+   */
+  EmbedFilterIntoGet();
+
+  /**
+   * Checks whether the given rule can be applied
+   * @param plan OperatorExpression to check
+   * @param context Current OptimizeContext executing under
+   * @returns Whether the input OperatorExpression passes the check
+   */
+  bool Check(common::ManagedPointer<OperatorExpression> plan, OptimizeContext *context) const override;
+
+  /**
+   * Transforms the input expression using the given rule
+   * @param input Input OperatorExpression to transform
+   * @param transformed Vector of transformed OperatorExpressions
+   * @param context Current OptimizeContext executing under
+   */
+  void Transform(common::ManagedPointer<OperatorExpression> input,
+                 std::vector<std::unique_ptr<OperatorExpression>> *transformed,
+                 OptimizeContext *context) const override;
+};
+
+///////////////////////////////////////////////////////////////////////////////
+/// Unnesting rules
+// We use this promise to determine which rules should be applied first if
+// multiple rules are applicable, we need to first pull filters up through mark-join
+// then turn mark-join into a regular join operator
+enum class UnnestPromise { Low = 1, High };
```

Comment:
This is weird having this inside of here. It should be moved somewhere else where it can be easily found.